### DISCUSSION TITLE
MUI mechanic question

### DISCUSSION THREAD

**Discussion by ViacheslavK9** [olikraus/u8g2#2349] [Q&A] [answered]
Hi!
Looking for better understanding what is the best way using MUI i found this:
in your example MUIBlink you have just a simple draw line function:

![image](https://github.com/olikraus/u8g2/assets/5678922/039d732f-aa36-40aa-bdc7-82fde0affde2)

I added two Serial.PrintLn lines to understand how is it working.
On a picture you can see my serial monitor with 8х2 lines - all i did - is just one Next click button.
So, one time next button = 8 times redraw? Could you say please, why is that? Thanks.. 

P.S. Continue enjoying your library, now playing with memory optimization, 
looking where is the right place for the right things.
(e.g. where to do calculations, heavy part on the MUI menu).

**Comment by olikraus** [comment, marked as answer]
> So, one time next button = 8 times redraw? Could you say please, why is that? 

Probably you have used the u8g2 page mode constructor (underscore 1 underscore).
If you would use the double page constructor (underscore 2 underscore) you will end up with only 4 redraws.
Using the full buffer constructor (underscore F underscore) will only require 1 redraw.

**Reply by ViacheslavK9** [reply]
thank you, full buffer takes much more program storage space (in terms of Arduino), 
so not an option, also i cant give up on mui, since i came from ERP field and user experience is important.

I use this constructor for now:
`//U8G2_SSD1306_128X64_NONAME_1_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE, /* clock=*/ 16, /* data=*/ 17);   // ESP32 Thing, HW I2C with pin remapping`

Not sure how i can switch to the "underscore 2 underscore" mode:
looked through all the list -- did not find anything like -could you help?

By the way i cut off some of the u8g2 parts by your article suggestions (rotating fonts and all that) - extra 3-4%! 👍

Also, could you tell me if "2 underscore" constructor could help me save memory?
thanks!

**Reply by olikraus** [reply]
> thank you, full buffer takes much more program storage space 

Thats the point: U8g2 allows you to trade memory consumption vs runtime... better runtime, more memory requirements ;-)

> Not sure how i can switch to the "underscore 2 underscore" mode:

simple, use `U8G2_SSD1306_128X64_NONAME_2_HW_I2C`

> could you tell me if "2 underscore" constructor could help me save memory?

It will not save memory, it will double speed at the cost of doubled frame buffer.

> parts by your article suggestions (rotating fonts and all that) - extra 3-4%!

fonts are the critical topic here. Try to reduce the size and the number of fonts. Use "r" or "n" fonts instead of "f" or "e" fonts. If nothing else helps, you could also start to create a custom fonts which exactly contain your glyphs which you need in your program (of course this requires a little bit of effort)

Also note, that other libs are often not written to safe memory. U8g2 and MUI are highly optimized for space. For example try to avoid the string class. It is probably worth to look at other libs too.
